libs/block-mesh-manager/src/ws/broadcaster.rs

```diff
@@ -0,0 +1,178 @@
+        count: usize,
+    ) -> usize {
+        let drained = self.move_queue(count);
+        let queued_messages_count = drained.len();
+        join_all(drained.into_iter().map(|user_id| {
+            let entry = self.sockets.get(&user_id).unwrap();
+            let tx = entry.value().clone();
+
+            let msgs = messages.clone();
+            async move {
+                for msg in msgs {
+                    if let Err(error) = tx.send(msg).await {
+                        tracing::error!("Error while queuing WS message: {error}");
+                    }
+                }
+            }
+        }))
+        .await;
+        queued_messages_count
+    }
+
+    pub fn subscribe(
+        &self,
+        user_id: Uuid,
+        socket_addr: SocketAddr,
+        sink_sender: mpsc::Sender<WsServerMessage>,
+    ) -> broadcast::Receiver<WsServerMessage> {
+        let old_value = self
+            .sockets
+            .insert((user_id, socket_addr), sink_sender.clone());
+        let queue = &mut self.queue.lock().unwrap();
+        queue.push_back((user_id, socket_addr));
+        debug_assert!(old_value.is_none());
+        self.global_transmitter.subscribe()
+    }
+
+    pub fn unsubscribe(&self, user_id: Uuid, socket_addr: SocketAddr) {
+        self.sockets.remove(&(user_id, socket_addr));
+        let queue = &mut self.queue.lock().unwrap();
+        if let Some(pos) = queue
+            .iter()
+            .position(|(a, b)| a == &user_id && b == &socket_addr)
+        {
+            queue.remove(pos);
+        } else {
+            tracing::error!("Failed to remove a socket from the queue");
+        }
+    }
+
+    pub async fn cron_reports(
+        &mut self,
+        period: Duration,
+        messages: impl Into<Vec<WsServerMessage>> + Clone + Send + 'static,
+        window_size: usize,
+        pool: PgPool,
+    ) -> anyhow::Result<CronReportsController> {
+        let (stats_tx, stats_rx) = tokio::sync::watch::channel(CronReportStats::new(
+            messages.clone().into(),
+            window_size,
+            0,
+            0,
+            Duration::from_secs(60),
+        ));
+        let broadcaster = self.clone();
+        let pool = pool.clone();
+        let user_id = Uuid::parse_str(
+            env::var(BLOCKMESH_SERVER_UUID_ENVAR)
+                .context("Could not find SERVER_UUID env var")?
+                .as_str(),
+        )
+        .context("SERVER_UUID evn var contains invalid UUID value")?;
+
+        tokio::spawn(async move {
+            let _ = settings_loop(
+                &pool,
+                &user_id,
+                period,
+                messages,
+                window_size,
+                broadcaster,
+                stats_tx,
+            )
+            .await;
+        });
+
+        let controller = CronReportsController::new(stats_rx);
+        self.cron_reports_controller = Some(controller.clone());
+        Ok(controller)
+    }
+}
```

Comment:
I think we can simplify this a lot.
Don't think we need an extra channel.
This entire function seems redundant, only the loop should exist.

Reply:
Yes, it's possible if we keep stats in the DB.